### PR TITLE
[fix] fixing failing randomized tests

### PR DESCRIPTION
**Pull request by akichidis**
## Description 

This PR is fixing an issue on the behaviour of `this_cert_parents_with_slow_nodes` which is used in randomized tests - and in other consensus unit tests as well. The method is picking the parents of a certificate supporting some failure modes (slow nodes). So basically , it does pick a subset of the given parents when slow nodes exist. However, it does not verifying that the parents honour the `2f + 1` requirement by stake. This is now fixed.

Now the randomized tests [should not fail](https://github.com/MystenLabs/sui/actions/runs/4793242232) anymore

## Test Plan 

Existing randomized_tests should not fail.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 1]
```diff
@@ -403,7 +404,11 @@ pub fn make_certificates_with_parameters(
                 parents_digests.insert(0, my_parent_digest);
             }
 
-            assert!(parents_digests.len() >= committee.quorum_threshold() as usize);
+            assert!(
+                parents_digests.len() >= committee.quorum_threshold() as usize,
+                "Failed on seed {}. At least 2f+1 parents are needed.",
+                seed
```
Comment (laura-makdah) [inline narwhal/consensus/src/tests/randomized_tests.rs:410]: Would this potentially cause flakiness where this sometimes fails? Can we instead re-generate the inputs if that is the case that some seeds fail?

**Review by laura-makdah — APPROVED** [review]

File: narwhal/consensus/src/tests/randomized_tests.rs (hunk 1 above)
Comment (akichidis) [reply narwhal/consensus/src/tests/randomized_tests.rs:410]: As we discussed offline the algorithm should always produce `2f+1` parents - if it doesn't then this is actually an issue/bug and this assert here ensures that we catch that.

File: narwhal/test-utils/src/lib.rs [hunk 2]
```diff
@@ -553,14 +565,35 @@ pub fn this_cert_parents_with_slow_nodes(
...
             // just add it directly as it is not within the slow nodes or we are the
             // same author.
             parents.insert(parent.digest());
+            total_stake += authority.stake();
         }
     }
 
+    // ensure we'll have enough parents (2f + 1)
+    while total_stake < committee.quorum_threshold() {
+        let parent = not_included.pop().unwrap();
```
Comment (mwtian) [inline narwhal/test-utils/src/lib.rs:582]: Is the order of `not_included` randomized?

**Review by mwtian — APPROVED** [review]

File: narwhal/test-utils/src/lib.rs (hunk 2 above)
Comment (akichidis) [reply narwhal/test-utils/src/lib.rs:582]: It isn't , but is not like a hard requirement to have it randomized - same as the original ancestors list, we don't randomise within this method before we process.

**Pull request merged**